### PR TITLE
chapter_13

### PR DESCRIPTION
**Pull request by chuan97sun**

### CODE REVIEW

**Comment by chuan97sun** [comment]
ready

File: app/assets/stylesheets/microposts.scss [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Microposts controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: https://sass-lang.com/
```
Comment (saobv-2223) [inline app/assets/stylesheets/microposts.scss]: remove file that never used

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -17,4 +17,12 @@ def extract_locale
     parsed_locale = params[:locale]
     I18n.available_locales.map(&:to_s).include?(parsed_locale) ? parsed_locale : nil
   end
+  
+   def logged_in_user
+    return if logged_in?
+
+      store_location
+      flash[:danger] = "please log in"
+      redirect_to login_path
```
Comment (saobv-2223) [inline app/controllers/application_controller.rb]:
```suggestion
    store_location
    flash[:danger] = "please log in"
    redirect_to login_path
```

File: app/controllers/microposts_controller.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
+class MicropostsController < ApplicationController
+  before_action :logged_in_user, only: [:create, :destroy]
+  before_action :correct_user, only: :destroy
+
+  def create
+    @micropost = current_user.microposts.build(micropost_params)
```
Comment (saobv-2223) [inline app/controllers/microposts_controller.rb]: remove ()

File: app/controllers/microposts_controller.rb [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+   def destroy
+     @micropost.destroy
+     flash[:success] = "Micropost deleted"
+     redirect_to request.referrer || root_url
+   end
+
+  private
+
+  def micropost_params
+    params.require(:micropost).permit(:content, :image)
```
Comment (saobv-2223) [inline app/controllers/microposts_controller.rb]: remove ()

File: app/helpers/microposts_helper.rb [hunk 5]
```diff
@@ -0,0 +1,2 @@
+module MicropostsHelper
+end
```
Comment (saobv-2223) [inline app/helpers/microposts_helper.rb]: remove file that never used

File: app/models/micropost.rb [hunk 6]
```diff
@@ -0,0 +1,12 @@
+class Micropost < ApplicationRecord
+  belongs_to :user
+  has_one_attached :image
+  default_scope -> { order(created_at: :desc) }
```
Comment (saobv-2223) [inline app/models/micropost.rb]: please, don't use default scope

File: app/models/user.rb [hunk 7]
```diff
@@ -68,6 +69,11 @@ def password_reset_expired?
     reset_sent_at < Settings.source.collection.reset_expired.hours.ago
   end
 
+  def feed
+    Micropost.where("user_id = ?", id)
+  end
+
+
```
Comment (saobv-2223) [inline app/models/user.rb]:
```suggestion
  def feed
    Micropost.where(user_id: id)
  end
  
```

File: app/views/microposts/_micropost.html.erb [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+  <span class="content">
+    <%= micropost.content %></span>
+    <%= image_tag micropost.image if micropost.image.attached? %>
+  <span class="timestamp">
+    Posted <%= time_ago_in_words(micropost.created_at) %> ago.
+    <% if current_user?(micropost.user) %>
+      <%= link_to "delete", micropost, method: :delete, data: { confirm: "You sure?" } %>
+    <% end %>
+  </span>
+</li>
```
Comment (saobv-2223) [inline app/views/microposts/_micropost.html.erb]:
```suggestion
</li>

```

File: app/views/shared/_feed.html.erb [hunk 9]
```diff
@@ -0,0 +1,6 @@
+<% if @feed_items.any? %>
+  <ol class="microposts">
+    <%= render @feed_items %>
+  </ol>
+  <%= paginate @feed_items, params: { controller: :static_page, action: :home } %>
+<% end %>
```
Comment (saobv-2223) [inline app/views/shared/_feed.html.erb]:
```suggestion
<% end %>

```

File: app/views/shared/_micropost_form.html.erb [hunk 10]
```diff
@@ -0,0 +1,19 @@
+<%= form_for(@micropost) do |f| %>
```
Comment (saobv-2223) [inline app/views/shared/_micropost_form.html.erb]: remove ()

File: app/views/shared/_micropost_form.html.erb [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+<% end %>
+
+<script type="text/javascript">
+  $("#micropost_image").bind("change", function() {
+    var size_in_megabytes = this.files[0].size/1024/1024;
+    if (size_in_megabytes > 5) {
+      alert("Maximum file size is 5MB. Please choose a smaller file.");
+    }
+  });
+</script>
```
Comment (saobv-2223) [inline app/views/shared/_micropost_form.html.erb]: please, do not write js in html file. Write in another .js file and call here

File: app/views/users/show.html.erb [hunk 12]
```diff
@@ -8,4 +8,13 @@
...
+   <div class="col-md-8">
+    <% if @user.microposts.any? %>
+      <h3><%= "Microposts" %> (<%= @user.microposts.count %>)</h3>
+      <ol class="microposts">
+        <%= render @microposts %>
+      </ol>
+      <%= paginate @microposts %>
+    <% end %>
+  </div>
+</div>
```
Comment (saobv-2223) [inline app/views/users/show.html.erb]:
```suggestion
</div>

```

**Comment by chuan97sun** [comment]
ready

**Comment by chuan97sun** [comment]
Ready

File: app/controllers/microposts_controller.rb [hunk 13]
```diff
@@ -0,0 +1,34 @@
+class MicropostsController < ApplicationController
+  before_action :logged_in_user, only: [:create, :destroy]
```
Comment (chungpv-1008) [inline app/controllers/microposts_controller.rb]:
```suggestion
  before_action :logged_in_user, only: %i(create destroy)
```

File: app/controllers/microposts_controller.rb [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+
+  private
+
+  def micropost_params
+    params.require(:micropost).permit :content, :image
+  end
+
+  def correct_user
+    @micropost = current_user.microposts.find_by id: params[:id]
+    redirect_to root_url if @micropost.nil?
```
Comment (chungpv-1008) [inline app/controllers/microposts_controller.rb]:
```suggestion
    redirect_to root_url if @micropost.blank?
```

File: app/models/user.rb [hunk 15]
```diff
@@ -68,6 +69,10 @@ def password_reset_expired?
     reset_sent_at < Settings.source.collection.reset_expired.hours.ago
   end
 
+  def feed
+    Micropost.where(user_id: id)
```
Comment (chungpv-1008) [inline app/models/user.rb]:
```suggestion
    microposts
```

File: app/views/microposts/_micropost.html.erb [hunk 16]
```diff
@@ -0,0 +1,13 @@
+<li id="micropost-<%= micropost.id %>">
+  <%= link_to gravatar_for(micropost.user, size: 50), micropost.user %>
+  <span class="user"><%= link_to micropost.user.name, micropost.user %></span>
+  <span class="content">
+    <%= micropost.content %></span>
+    <%= image_tag micropost.image if micropost.image.attached? %>
+  <span class="timestamp">
+    Posted <%= time_ago_in_words(micropost.created_at) %> ago.
+    <% if current_user?(micropost.user) %>
```
Comment (chungpv-1008) [inline app/views/microposts/_micropost.html.erb]:
```suggestion
    Posted <%= time_ago_in_words micropost.created_at %> ago.
    <% if current_user? micropost.user %>
```

File: app/views/shared/_error_messages.html.erb [hunk 17]
```diff
@@ -1,10 +1,10 @@
-<% if @user.errors.any? %>
+<% if object.errors.any? %>
   <div id="error_explanation">
     <div class="alert alert-danger">
-      The form contains <%= pluralize @user.errors.size, "error" %>.
+      The form contains <%= pluralize(object.errors.count, "error") %>.
```
Comment (chungpv-1008) [inline app/views/shared/_error_messages.html.erb]:
```suggestion
      The form contains <%= pluralize(object.errors.size, "error") %>.
```

File: app/views/shared/_micropost_form.html.erb [hunk 18]
```diff
@@ -0,0 +1,10 @@
+<%= form_for @micropost do |f| %>
+  <%= render 'shared/error_messages', object: f.object %>
```
Comment (chungpv-1008) [inline app/views/shared/_micropost_form.html.erb]: using " instead of '

File: app/views/shared/_user_info.html.erb [hunk 19]
```diff
@@ -0,0 +1,4 @@
+<%= link_to gravatar_for(current_user, size: 50), current_user %>
+<h1><%= current_user.name %></h1>
+<span><%= link_to "view my profile", current_user %></span>
+<span><%= pluralize(current_user.microposts.count, "micropost") %></span>
```
Comment (chungpv-1008) [inline app/views/shared/_user_info.html.erb]:
```suggestion
<span><%= pluralize current_user.microposts.size, "micropost" %></span>
```

File: app/views/users/show.html.erb [hunk 20]
```diff
@@ -8,4 +8,13 @@
       </h1>
     </section>
   </aside>
+   <div class="col-md-8">
+    <% if @user.microposts.any? %>
+      <h3><%= "Microposts" %> (<%= @user.microposts.count %>)</h3>
```
Comment (chungpv-1008) [inline app/views/users/show.html.erb]:
```suggestion
      <h3><%= "Microposts" %> (<%= @user.microposts.size %>)</h3>
```

**Comment by chuan97sun** [comment]
ready

**Comment by vanvtt-0952** [comment]
merged

**Pull request merged**